### ISSUE TITLE
Potential to reuse validation policies in the Datree project

### ISSUE THREAD

**Issue by taham0** [xlab-uiuc/acto#191]
**TL;DR: The built-in policies of the Datree tool might not be very helpful for Acto because violations of many policies does not result in state inconsistency or unhealthy resource. Also, its yaml interface for writing custom policies might not be as flexible as tools like [Open Policy Agent (OPA)](https://www.openpolicyagent.org) that utilizes "policy-as-code".** 

So @Essoz and I have looked at the Datree tool together in the hope to understand better the benefit of reusing it in the acto project.
Our expectation: The Datree tool comes with a set of advanced policies that can be used to validate the kubernetes manifest. We hope to reuse these policies to help Acto understand the semantic correctness of its generated input better and find more bugs.

We inspected and categorized Datree’s built-in rules ([link to all rules](https://hub.datree.io/built-in-rules)) to see how helpful they are.

![image](https://user-images.githubusercontent.com/76549451/199728630-d22b4a26-7a9f-4b89-a2b0-34d28f382897.png)

# Categories

### Unhealthy Resource (3 rules)
Rules in this category prevents misconfigurations that eventually lead to failures during rolling out from entering the cluster. For [example](https://hub.datree.io/built-in-rules/ensure-valid-restart-policy), having an not-allowed option in `spec.template.spec.restartPolicy` will lead to pod creation failures.

### Security (8 rules)
This category encompasses all such built-in rules that are concerned with preventing security issues due to misconfigurations in kubernetes manifests. For example, [this rule](https://hub.datree.io/built-in-rules/prevent-node-port) addresses the NodePort service type and prevents it from being set. This is a security concern because it opens a network port on each node that makes all of them accessible to the cluster’s external network.

### Maintenance (9 rules)
This category includes all built-in rules that prevent any such configurations that could cause problems in how the cluster is maintained. For example, if a raw pod resource is configured as part of the manifest, [this rule](https://hub.datree.io/built-in-rules/prevent-naked-pods) is violated because it is preferable to use higher-level resources within clusters.

### Performance (2 rules)
This category includes built-in rules that prevent configurations that could lead to performance issues within the cluster. For example, specifying the ingress host to a wildcard character (*) can lead to all traffic being forwarded to single node as described [here](https://hub.datree.io/built-in-rules/prevent-ingress-forwarding-traffic-to-single-container).

### Unhealthy Resource Under Specific Conditions (12 rules)
Rules in this category prevents potential resiliency issues under certain conditions. For [example](https://hub.datree.io/built-in-rules/ensure-memory-request),  not setting resource requests and limits can lead to undefined behaviors in scheduler and cluster autoscaler. Pods can also get OOMKilled due to resource contention.

From our discussions, only *Unhealthy Resource* matters to Acto, since only this category lead to failures of resource update/creation which prevents the operator from reconciling system state. Other categories can be less important to Acto because they either require time or specific environment (e.g. an overloaded cluster) to manifest or do not lead to state inconsistencies. 

As a result, we suspect the built-in rules in Datree are not very helpful to Acto since unhealthy resource only has three rules.

# Feasibility of Writing Custom Policies
Users can write custom policies using JSON or YAML schema. It is a long discussed limitation that YAML or JSON schemas cannot express the relationship between multiple fields very well. Though it is possible to use conditions in these schemas, the expressiveness is still very limited as compared to higher level programming language. Indeed, a popular way to write custom validation policies is to use [Open Policy Agent (OPA)](https://www.openpolicyagent.org), which is a CNCF graduated project and used by many K8S validation and monitoring tools like [Kubeeye](https://github.com/kubesphere/kubeeye)

Hence, if we were to enhance Acto as the expectation, we think using a tool like OPA will be a better choice than Datree.
